### PR TITLE
Support attrs decorators even if they are imported from attrs

### PR DESCRIPTION
**Pull request by alonme**
This PR is a POC for improving the way that astroid detects `attrs` decorators.
I am assuming that adding `frozen`, `define`, and `mutable` to `ATTRS_NAMES` was considered and was decided against - otherwise, it should probably be preferred

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes https://github.com/PyCQA/pylint/issues/7884

### CODE REVIEW

**Comment by DanielNoord** [comment]
@alonme Sorry for pushing to your branch but I think this might be a better solution. Instead of using some hack we actually use our inference brain.

@Pierre-Sassoulas Do we have a good way to measure the performance impact of this change?

**Comment by Pierre-Sassoulas** [comment]
No we'd need a performance primer, this has been on my mind for quite some time now but there is always something else to do :)

**Comment by alonme** [comment]
@DanielNoord  - No worries, this is of course much better.

Tell me if i can help with anything else

**Comment by DanielNoord** [comment]
> @DanielNoord - No worries, this is of course much better.
> 
> Tell me if i can help with anything else

Well, I'm not sure if you have a large codebase at work or for personal projects but it would be really good to see how long `pylint` takes with this PR merged into `astroid` for a project that currently takes comparatively long to lint already. 

Oh, and we need a changelog entry :smile:

**Comment by Pierre-Sassoulas** [comment]
We do have a benchmark test in pylint, it's not automated but you can check the difference between two versions of editable install of astroid. clone pylint then pip install -e your local astroid, launch ``pytest -k benchmark`` once for main astroid, once for this branch and compare the result.

**Comment by alonme** [comment]
Original

```
---------------------------------------------------------------------------------------------------------------- benchmark 'baseline': 11 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                  Min                       Max                      Mean                 StdDev                    Median                    IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_baseline_benchmark_j1                                    642.9770 (1.0)          1,428.7080 (1.0)            739.2487 (1.0)         112.0223 (1.0)            694.5560 (1.0)         101.8795 (1.0)         89;53  1,352.7248 (1.0)         613           1
test_baseline_benchmark_j1_all_checks_single_file         116,752.9580 (181.58)     211,995.9280 (148.38)     141,993.9449 (192.08)   29,315.4609 (261.69)     136,810.1150 (196.97)   11,071.6855 (108.67)        1;1      7.0426 (0.01)          8           1
test_baseline_lots_of_files_j1_empty_checker              223,609.7690 (347.77)     232,891.4640 (163.01)     228,809.2736 (309.52)    3,556.4965 (31.75)      228,393.3320 (328.83)    4,818.5200 (47.30)         2;0      4.3705 (0.00)          5           1
test_baseline_lots_of_files_j1                            225,234.3440 (350.30)     280,985.0320 (196.67)     241,983.4922 (327.34)   22,436.6740 (200.29)     234,849.1460 (338.13)   21,192.6762 (208.02)        1;0      4.1325 (0.00)          5           1
test_baseline_benchmark_j1_all_checks_lots_of_files       242,627.6570 (377.35)     261,145.6920 (182.78)     252,308.4546 (341.30)    7,917.0373 (70.67)      249,653.8560 (359.44)   13,229.2232 (129.85)        2;0      3.9634 (0.00)          5           1
test_baseline_benchmark_j1_single_working_checker         504,609.7670 (784.80)     508,595.6660 (355.98)     506,773.7986 (685.53)    1,528.7661 (13.65)      506,435.9390 (729.15)    2,057.6145 (20.20)         2;0      1.9733 (0.00)          5           1
test_baseline_benchmark_check_parallel_j2                 860,153.1770 (>1000.0)    873,649.6280 (611.50)     867,640.9618 (>1000.0)   5,228.4055 (46.67)      868,421.8980 (>1000.0)   7,659.0113 (75.18)         2;0      1.1526 (0.00)          5           1
test_baseline_benchmark_j2                                861,621.1170 (>1000.0)    886,040.1400 (620.17)     869,244.8098 (>1000.0)   9,799.4752 (87.48)      866,038.2450 (>1000.0)  10,326.3895 (101.36)        1;0      1.1504 (0.00)          5           1
test_baseline_lots_of_files_j2                          1,095,141.9380 (>1000.0)  1,109,569.9160 (776.62)   1,101,237.5432 (>1000.0)   5,580.7763 (49.82)    1,099,350.8350 (>1000.0)   7,628.3482 (74.88)         2;0      0.9081 (0.00)          5           1
test_baseline_lots_of_files_j2_empty_checker            1,153,549.8460 (>1000.0)  1,235,064.6380 (864.46)   1,195,213.4776 (>1000.0)  34,162.0979 (304.96)   1,201,637.4480 (>1000.0)  58,508.2465 (574.29)        2;0      0.8367 (0.00)          5           1
test_baseline_benchmark_j2_single_working_checker       1,396,959.8420 (>1000.0)  1,442,936.7230 (>1000.0)  1,412,449.0406 (>1000.0)  18,095.8019 (161.54)   1,405,713.4690 (>1000.0)  19,772.2875 (194.08)        1;0      0.7080 (0.00)          5           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

New Astroid
```
---------------------------------------------------------------------------------------------------------------- benchmark 'baseline': 11 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                  Min                       Max                      Mean                 StdDev                    Median                    IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_baseline_benchmark_j1                                    641.5790 (1.0)          1,328.2370 (1.0)            736.4457 (1.0)         106.5480 (1.0)            691.5840 (1.0)          95.7635 (1.0)         69;49  1,357.8734 (1.0)         473           1
test_baseline_benchmark_j1_all_checks_single_file         111,243.2120 (173.39)     189,752.4330 (142.86)     122,381.1555 (166.18)   27,267.7914 (255.92)     112,297.6815 (162.38)    3,274.8270 (34.20)         1;1      8.1712 (0.01)          8           1
test_baseline_lots_of_files_j1_empty_checker              224,901.3490 (350.54)     246,673.0210 (185.71)     235,200.4922 (319.37)   10,376.5844 (97.39)      234,226.3730 (338.68)   20,141.1570 (210.32)        1;0      4.2517 (0.00)          5           1
test_baseline_lots_of_files_j1                            226,492.3970 (353.02)     233,275.2050 (175.63)     228,564.7700 (310.36)    2,752.9989 (25.84)      228,071.7350 (329.78)    2,900.1143 (30.28)         1;0      4.3751 (0.00)          5           1
test_baseline_benchmark_j1_all_checks_lots_of_files       231,990.3950 (361.59)     241,504.4150 (181.82)     236,295.9408 (320.86)    4,392.1657 (41.22)      235,916.5710 (341.12)    8,304.6982 (86.72)         1;0      4.2320 (0.00)          5           1
test_baseline_benchmark_j1_single_working_checker         505,705.3120 (788.22)     510,921.8300 (384.66)     509,039.6542 (691.21)    2,207.7081 (20.72)      510,081.9000 (737.56)    3,334.7282 (34.82)         1;0      1.9645 (0.00)          5           1
test_baseline_benchmark_check_parallel_j2                 856,387.9930 (>1000.0)    880,670.6260 (663.04)     866,981.7080 (>1000.0)   8,974.2807 (84.23)      865,482.9950 (>1000.0)  10,696.0433 (111.69)        2;0      1.1534 (0.00)          5           1
test_baseline_benchmark_j2                                857,057.6400 (>1000.0)    916,018.9590 (689.65)     873,404.7904 (>1000.0)  24,171.7238 (226.86)     865,202.7050 (>1000.0)  19,876.2832 (207.56)        1;1      1.1449 (0.00)          5           1
test_baseline_lots_of_files_j2                          1,079,281.9450 (>1000.0)  1,105,041.7630 (831.96)   1,094,832.0024 (>1000.0)   9,925.2562 (93.15)    1,098,884.7370 (>1000.0)  12,321.1958 (128.66)        2;0      0.9134 (0.00)          5           1
test_baseline_lots_of_files_j2_empty_checker            1,102,761.4370 (>1000.0)  1,121,359.2350 (844.25)   1,110,383.5888 (>1000.0)   8,500.3720 (79.78)    1,107,454.7770 (>1000.0)  15,350.8312 (160.30)        1;0      0.9006 (0.00)          5           1
test_baseline_benchmark_j2_single_working_checker       1,391,254.1400 (>1000.0)  1,403,989.3820 (>1000.0)  1,396,174.4972 (>1000.0)   5,111.8423 (47.98)    1,396,403.4950 (>1000.0)   7,196.2392 (75.15)         1;0      0.7162 (0.00)          5           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I'll let you be the judge. Is that good enough?

**Comment by alonme** [comment]
Is there a reason why we run it with `warmup = False`?

**Comment by Pierre-Sassoulas** [comment]
> Is there a reason why we run it with warmup = False?

Probably in order for the tests to be a little faster. To be fair we're never actually looking at the result (and abnormally slow result do not fail) so we may as well remove them from the pylint pipeline until we make it better. Making it better is probably the first step if we want to improve performance in pylint though.

It doesn't seem the performance are that much worse right now, seeing the benchmark it even seems it's faster (?).

**Comment by DanielNoord** [comment]
> It doesn't seem the performance are that much worse right now, seeing the benchmark it even seems it's faster (?).

That makes me more worried about the usefulness of the test than give me confidence in the change :sweat_smile:

**Comment by alonme** [comment]
Yes, i saw that, but given that it was the second test i ran, i thought the missing warmup could have been part of the issue

Anyway, ran it again with warmup (this time new-astroid was run first).

Results:

Original
```
---------------------------------------------------------------------------------------------------------------- benchmark 'baseline': 11 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                  Min                       Max                      Mean                 StdDev                    Median                    IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_baseline_benchmark_j1                                    643.8640 (1.0)          1,535.2660 (1.0)            702.3279 (1.0)         104.5348 (1.0)            663.1580 (1.0)          37.6168 (1.0)       156;218  1,423.8364 (1.0)        1557           1
test_baseline_benchmark_j1_all_checks_single_file         110,553.8280 (171.70)     196,276.4300 (127.85)     123,253.4510 (175.49)   25,964.4960 (248.38)     114,671.6445 (172.92)    3,275.4880 (87.08)         1;2      8.1134 (0.01)         10           1
test_baseline_lots_of_files_j1_empty_checker              217,733.7290 (338.17)     222,033.0610 (144.62)     220,264.1880 (313.62)    1,691.5706 (16.18)      221,021.7090 (333.29)    2,302.4817 (61.21)         2;0      4.5400 (0.00)          5           1
test_baseline_lots_of_files_j1                            220,893.3910 (343.07)     267,561.8290 (174.28)     232,549.0864 (331.11)   19,673.7925 (188.20)     225,591.3020 (340.18)   13,681.3545 (363.70)        1;1      4.3002 (0.00)          5           1
test_baseline_benchmark_j1_all_checks_lots_of_files       224,891.8040 (349.28)     231,997.2170 (151.11)     229,170.9508 (326.30)    2,729.9111 (26.11)      229,778.5570 (346.49)    3,546.2108 (94.27)         2;0      4.3636 (0.00)          5           1
test_baseline_benchmark_j1_single_working_checker         508,633.2360 (789.97)     510,160.0100 (332.29)     509,481.3882 (725.42)      768.2406 (7.35)       509,892.1530 (768.88)    1,433.3315 (38.10)         2;0      1.9628 (0.00)          5           1
test_baseline_benchmark_j2                                819,943.0350 (>1000.0)    841,147.4740 (547.88)     825,922.7904 (>1000.0)   8,811.8620 (84.30)      821,569.0370 (>1000.0)   8,995.2258 (239.13)        1;0      1.2108 (0.00)          5           1
test_baseline_benchmark_check_parallel_j2                 823,139.1360 (>1000.0)    841,696.9730 (548.24)     833,274.8462 (>1000.0)   6,849.4428 (65.52)      833,810.0580 (>1000.0)   8,465.7672 (225.05)        2;0      1.2001 (0.00)          5           1
test_baseline_lots_of_files_j2                          1,043,067.6570 (>1000.0)  1,070,368.7580 (697.19)   1,051,649.5148 (>1000.0)  10,860.7232 (103.90)   1,049,494.2460 (>1000.0)  10,384.5150 (276.06)        1;0      0.9509 (0.00)          5           1
test_baseline_lots_of_files_j2_empty_checker            1,060,807.9360 (>1000.0)  1,078,688.6390 (702.61)   1,067,107.3374 (>1000.0)   6,959.8230 (66.58)    1,066,083.7050 (>1000.0)   7,796.2030 (207.25)        1;0      0.9371 (0.00)          5           1
test_baseline_benchmark_j2_single_working_checker       1,353,542.5310 (>1000.0)  1,440,143.0620 (938.04)   1,384,070.7768 (>1000.0)  35,816.7290 (342.63)   1,364,300.3410 (>1000.0)  48,644.7148 (>1000.0)       1;0      0.7225 (0.00)          5           1
---------------------------------------------------------------------------------
```

New Astroid
```
---------------------------------------------------------------------------------------------------------------- benchmark 'baseline': 11 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                  Min                       Max                      Mean                 StdDev                    Median                    IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_baseline_benchmark_j1                                    643.6800 (1.0)          1,385.7650 (1.0)            699.6720 (1.0)          93.4275 (1.0)            662.3530 (1.0)          37.8915 (1.0)       170;219  1,429.2410 (1.0)        1556           1
test_baseline_benchmark_j1_all_checks_single_file         107,613.7160 (167.19)     189,238.1060 (136.56)     118,101.5733 (168.80)   25,064.6276 (268.28)     110,426.1220 (166.72)    3,655.7380 (96.48)         1;1      8.4673 (0.01)         10           1
test_baseline_lots_of_files_j1                            217,064.9380 (337.22)     221,443.4810 (159.80)     218,977.2614 (312.97)    1,736.0008 (18.58)      218,727.2380 (330.23)    2,670.3123 (70.47)         2;0      4.5667 (0.00)          5           1
test_baseline_lots_of_files_j1_empty_checker              218,505.6130 (339.46)     222,641.1000 (160.66)     220,896.7388 (315.71)    1,517.3827 (16.24)      221,162.3180 (333.90)    1,620.9245 (42.78)         2;0      4.5270 (0.00)          5           1
test_baseline_benchmark_j1_all_checks_lots_of_files       228,018.5380 (354.24)     232,593.9540 (167.85)     230,751.3178 (329.80)    1,744.3109 (18.67)      231,152.4640 (348.99)    2,218.5342 (58.55)         2;0      4.3337 (0.00)          5           1
test_baseline_benchmark_j1_single_working_checker         507,398.0170 (788.28)     510,915.2930 (368.69)     508,767.2706 (727.15)    1,679.9224 (17.98)      507,714.0800 (766.53)    2,906.4948 (76.71)         1;0      1.9655 (0.00)          5           1
test_baseline_benchmark_check_parallel_j2                 818,285.0140 (>1000.0)    843,333.7000 (608.57)     827,577.7448 (>1000.0)   9,512.0189 (101.81)     823,941.6370 (>1000.0)   9,631.5788 (254.19)        1;0      1.2083 (0.00)          5           1
test_baseline_benchmark_j2                                819,661.3150 (>1000.0)    848,752.0630 (612.48)     827,964.9502 (>1000.0)  12,007.9426 (128.53)     824,943.3120 (>1000.0)  12,226.6432 (322.68)        1;0      1.2078 (0.00)          5           1
test_baseline_lots_of_files_j2                          1,044,696.8820 (>1000.0)  1,166,459.6680 (841.74)   1,098,367.1356 (>1000.0)  53,764.6922 (575.47)   1,093,277.5160 (>1000.0)  97,321.4050 (>1000.0)       1;0      0.9104 (0.00)          5           1
test_baseline_lots_of_files_j2_empty_checker            1,058,479.1550 (>1000.0)  1,071,314.0610 (773.08)   1,065,996.4548 (>1000.0)   4,696.0535 (50.26)    1,066,967.6080 (>1000.0)   4,602.6570 (121.47)        2;0      0.9381 (0.00)          5           1
test_baseline_benchmark_j2_single_working_checker       1,355,346.3720 (>1000.0)  1,361,618.1500 (982.58)   1,358,613.4226 (>1000.0)   2,389.8686 (25.58)    1,359,100.9520 (>1000.0)   3,368.2063 (88.89)         2;0      0.7360 (0.00)          5           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```


These results seemed to make more sensed, and then i figured that it contains other astroid changes beside this PR, so i ran the benchmark on the main branch.

Running on `main` - 6714f155


```
---------------------------------------------------------------------------------------------------------------- benchmark 'baseline': 11 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                  Min                       Max                      Mean                 StdDev                    Median                    IQR            Outliers         OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_baseline_benchmark_j1                                    636.6540 (1.0)          1,787.2520 (1.0)            700.0577 (1.0)         112.5502 (1.0)            658.5395 (1.0)          41.2580 (1.0)       158;225  1,428.4536 (1.0)        1578           1
test_baseline_benchmark_j1_all_checks_single_file         107,652.6820 (169.09)     197,647.5740 (110.59)     118,124.8781 (168.74)   27,963.5870 (248.45)     109,224.3950 (165.86)    1,915.2700 (46.42)         1;1      8.4656 (0.01)         10           1
test_baseline_lots_of_files_j1_empty_checker              216,989.0830 (340.83)     222,075.2390 (124.26)     218,998.1260 (312.83)    2,064.0566 (18.34)      218,162.7760 (331.28)    3,050.0200 (73.93)         1;0      4.5662 (0.00)          5           1
test_baseline_lots_of_files_j1                            217,382.3460 (341.45)     221,370.7870 (123.86)     218,737.0778 (312.46)    1,627.2654 (14.46)      218,010.1570 (331.05)    2,129.3125 (51.61)         1;0      4.5717 (0.00)          5           1
test_baseline_benchmark_j1_all_checks_lots_of_files       223,605.0000 (351.22)     247,176.7990 (138.30)     230,382.9940 (329.09)    9,551.6216 (84.87)      226,421.3230 (343.82)    7,711.5002 (186.91)        1;1      4.3406 (0.00)          5           1
test_baseline_benchmark_j1_single_working_checker         506,644.4820 (795.79)     511,114.1390 (285.98)     508,515.7892 (726.39)    1,686.4206 (14.98)      507,902.3370 (771.26)    2,029.8718 (49.20)         2;0      1.9665 (0.00)          5           1
test_baseline_benchmark_check_parallel_j2                 816,301.4440 (>1000.0)    837,751.5110 (468.74)     824,796.8618 (>1000.0)   8,709.0381 (77.38)      821,769.9310 (>1000.0)  13,099.2505 (317.50)        1;0      1.2124 (0.00)          5           1
test_baseline_benchmark_j2                                818,502.0760 (>1000.0)    840,418.0930 (470.23)     825,982.6572 (>1000.0)   8,967.1939 (79.67)      824,483.5950 (>1000.0)  12,201.5603 (295.74)        1;0      1.2107 (0.00)          5           1
test_baseline_lots_of_files_j2                          1,037,724.2520 (>1000.0)  1,054,704.4680 (590.13)   1,047,569.3632 (>1000.0)   6,597.7499 (58.62)    1,047,037.1500 (>1000.0)   9,076.1295 (219.98)        2;0      0.9546 (0.00)          5           1
test_baseline_lots_of_files_j2_empty_checker            1,064,759.4460 (>1000.0)  1,070,382.6140 (598.90)   1,067,297.6558 (>1000.0)   2,470.0124 (21.95)    1,066,371.2360 (>1000.0)   4,313.2473 (104.54)        2;0      0.9369 (0.00)          5           1
test_baseline_benchmark_j2_single_working_checker       1,357,209.2300 (>1000.0)  1,462,233.6560 (818.15)   1,382,995.9402 (>1000.0)  44,662.4133 (396.82)   1,363,501.4780 (>1000.0)  35,671.3478 (864.59)        1;1      0.7231 (0.00)          5           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```


To me, it doesn't look like a big performance hit, considering these results and the run on the pandas codebase.

**Comment by alonme** [comment]
Pushed some more cases to the tests and added a changelog

**Review by Pierre-Sassoulas — APPROVED** [review]

File: tests/brain/test_attr.py [hunk 1]
```diff
@@ -141,12 +141,31 @@ class Eggs:
...
 
-        for name in ("f", "g", "h", "i", "j", "k", "l"):
+        for name in ("f", "g", "h", "i", "j", "k", "l", "m", "n"):
             should_be_unknown = next(module.getattr(name)[0].infer()).getattr("d")[0]
-            self.assertIsInstance(should_be_unknown, astroid.Unknown)
+            try:
+                self.assertIsInstance(should_be_unknown, astroid.Unknown)
+            except:
+                # breakpoint()
+                ...
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply tests/brain/test_attr.py]: (Unpublished review on an old diff, resolving)

**Comment by alonme** [comment]
@Pierre-Sassoulas @DanielNoord thanks for the help!

Should i open an issue in Pylint to check the benchmark issues? i i think that the warmup may be important

**Comment by Pierre-Sassoulas** [comment]
Sounds good, we could also use https://pytest-benchmark.readthedocs.io/en/latest/comparing.html#comparing-past-runs if you feel brave enough to do it 😄

**Pull request merged**